src/core/DesktopManager.ts

```diff
@@ -0,0 +1,673 @@
+import { LayoutManager, Monitor } from "@schnz/gnome-shell/src/ui/layout.js";
+
+import Meta from "gi://Meta?version=13";
+import Mtk from "gi://Mtk?version=13";
+
+import { DesktopEvent, Event } from "../types/desktop.js";
+import {
+  GridOffset,
+  GridSelection,
+  GridSize,
+  Rectangle
+} from "../types/grid.js";
+import { CardinalDirection } from "../types/hotkeys.js";
+import { DispatchFn, Publisher } from "../types/observable.js";
+import { Node } from "../types/tree.js";
+import { GarbageCollection, GarbageCollector } from "../util/gc.js";
+import { adjust, pan } from "../util/grid.js";
+import { GridSpec } from "../util/parser.js";
+import { UserPreferencesProvider } from "./UserPreferences.js";
+
+// splits computed gridspec cell areas in non-dynamic and dynamic cells
+type GridSpecAreas = [dedicated: Rectangle[], dynamic: Rectangle[]];
+
+type FrameSize = { width: number; height: number };
+
+type MRect = Mtk.Rectangle;
+
+export interface DesktopManagerParams {
+  display: Meta.Display;
+  layoutManager: LayoutManager;
+  monitorManager: Meta.MonitorManager;
+  workspaceManager: Meta.WorkspaceManager;
+  userPreferences: UserPreferencesProvider;
+}
+
+/**
+ * Abstracts over a multitude of Gnome APIs to provide a unified interface for
+ * desktop-related actions and window manipulation.
+ */
+export default class implements Publisher<DesktopEvent>, GarbageCollector {
+  #gc: GarbageCollection;
+  #display: Meta.Display;
+  #layoutManager: LayoutManager;
+  #workspaceManager: Meta.WorkspaceManager;
+  #userPreferences: UserPreferencesProvider;
+  #dispatchCallbacks: DispatchFn<DesktopEvent>[];
```

Comment:
Whats up with the `#` prefixes are you using those to denote they should be private?

Reply:
I did some googling and found that apparently in vanilla javascript you can use that to declare visability.

However this isn't idiomatic in Typescript, you can just use the `private` keyword

```typescript
private gc: GarbageColleciton
```

Reply:
I beg to differ. It's native ECMAScript which provides _native_ private members for classes. Just because its a more recent feature and hasn't yet been adapted in TypeScript projects doesn't mean that it isn't idiomatic.